File: frontend/js/views/leverage_dol_dfl.js
// Degree of leverage — operating (DOL), financial (DFL), and combined (DCL),
// via /calc/leverage. Live as you type.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';

const money = (n) => '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 0, maximumFractionDigits: 0 });
const x = (n) => (n == null ? '—' : Number(n).toLocaleString(undefined, { maximumFractionDigits: 3 }) + '×');

export async function renderLeverage(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.leverage.h1.title">// DEGREE OF LEVERAGE</span></h1>
        <p class="muted small" data-i18n="view.leverage.hint.intro">
            How a change in sales is amplified down the income statement. Fixed operating costs
            amplify it into EBIT (operating leverage); fixed interest amplifies it again into EPS
            (financial leverage). A DCL of 3 means a 1% change in sales swings EPS 3%. Updates as
            you type.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.leverage.h2.inputs">The income statement</h2>
            <form id="leverage-form" class="inline-form">
                <label><span data-i18n="view.leverage.label.sales">Sales ($)</span>
                    <input type="number" step="0.01" name="sales_usd" value="1000" required></label>
                <label><span data-i18n="view.leverage.label.variable">Variable costs ($)</span>
                    <input type="number" step="0.01" min="0" name="variable_costs_usd" value="400" required></label>
                <label><span data-i18n="view.leverage.label.fixed">Fixed costs ($)</span>
                    <input type="number" step="0.01" min="0" name="fixed_costs_usd" value="300" required></label>
                <label><span data-i18n="view.leverage.label.interest">Interest expense ($)</span>
                    <input type="number" step="0.01" min="0" name="interest_expense_usd" value="100"></label>
            </form>
        </div>
        <div id="leverage-result" class="lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#leverage-form');
    const generate = async () => {
        const fd = new FormData(form);
        const body = {
            sales_usd: Number(fd.get('sales_usd')) || 0,
            variable_costs_usd: Number(fd.get('variable_costs_usd')) || 0,
            fixed_costs_usd: Number(fd.get('fixed_costs_usd')) || 0,
            interest_expense_usd: Number(fd.get('interest_expense_usd')) || 0,
        };
        try {
            const r = await api.calcLeverage(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, r);
        } catch (err) {
            showToast(err.message || t('view.leverage.toast.error'), { level: 'error' });
        }
    };
    form.addEventListener('input', debounce(generate, 250));
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function renderResult(mount, r) {
    const el = mount.querySelector('#leverage-result');
    el.innerHTML = `
        <div class="chart-panel">
            <h2 data-i18n="view.leverage.h2.result">The amplification</h2>
            <div class="cards">
                <div class="card"><div class="label" data-i18n="view.leverage.card.dol">Operating (DOL)</div>
                    <div class="value">${x(r.dol)}</div></div>
                <div class="card"><div class="label" data-i18n="view.leverage.card.dfl">Financial (DFL)</div>
                    <div class="value">${x(r.dfl)}</div></div>
                <div class="card pos"><div class="label" data-i18n="view.leverage.card.dcl">Combined (DCL)</div>
                    <div class="value pos">${x(r.dcl)}</div></div>
            </div>
            <table class="data-table">
                <tbody>
                    <tr><td data-i18n="view.leverage.row.cm">Contribution margin</td><td>${money(r.contribution_margin_usd)}</td></tr>
                    <tr><td data-i18n="view.leverage.row.ebit">EBIT</td><td>${money(r.ebit_usd)}</td></tr>
                    <tr><td data-i18n="view.leverage.row.pretax">Pre-tax income</td><td>${money(r.pretax_income_usd)}</td></tr>
                    <tr><td data-i18n="view.leverage.row.dol">Operating leverage (DOL)</td><td>${x(r.dol)}</td></tr>
                    <tr><td data-i18n="view.leverage.row.dfl">Financial leverage (DFL)</td><td>${x(r.dfl)}</td></tr>
                    <tr class="emph"><td data-i18n="view.leverage.row.dcl">Combined leverage (DCL)</td><td>${x(r.dcl)}</td></tr>
                </tbody>
            </table>
        </div>
    `;
    applyUiI18n(el);
}

File: crates/traderview-core/src/leverage_dol_dfl.rs
//! Degree of operating, financial, and combined leverage.
//!
//! Leverage measures how a change in sales is amplified on the way down the
//! income statement — fixed operating costs amplify it into EBIT, fixed
//! interest amplifies it again into EPS.
//!
//! ```text
//! contribution margin = sales − variable costs
//! EBIT                = contribution margin − fixed costs
//! DOL = contribution margin / EBIT          (sales → EBIT amplification)
//! DFL = EBIT / (EBIT − interest)            (EBIT → EPS amplification)
//! DCL = DOL × DFL = contribution margin / (EBIT − interest)
//! ```
//!
//! A DOL of 2 means a 1% change in sales swings EBIT 2%; DCL chains both, so a
//! 1% sales change swings EPS by DCL%.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct LeverageInput {
    pub sales_usd: f64,
    pub variable_costs_usd: f64,
    pub fixed_costs_usd: f64,
    #[serde(default)]
    pub interest_expense_usd: f64,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct LeverageResult {
    /// sales − variable costs.
    pub contribution_margin_usd: f64,
    /// contribution margin − fixed costs.
    pub ebit_usd: f64,
    /// Pre-tax income: EBIT − interest.
    pub pretax_income_usd: f64,
    /// Degree of operating leverage; `None` if EBIT is 0.
    pub dol: Option<f64>,
    /// Degree of financial leverage; `None` if pre-tax income is 0.
    pub dfl: Option<f64>,
    /// Degree of combined leverage; `None` if pre-tax income is 0.
    pub dcl: Option<f64>,
}

pub fn analyze(input: &LeverageInput) -> LeverageResult {
    let cm = input.sales_usd - input.variable_costs_usd;
    let ebit = cm - input.fixed_costs_usd;
    let pretax = ebit - input.interest_expense_usd;

    let dol = if ebit != 0.0 { Some(cm / ebit) } else { None };
    let dfl = if pretax != 0.0 { Some(ebit / pretax) } else { None };
    let dcl = if pretax != 0.0 { Some(cm / pretax) } else { None };

    LeverageResult {
        contribution_margin_usd: cm,
        ebit_usd: ebit,
        pretax_income_usd: pretax,
        dol,
        dfl,
        dcl,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 1e-6
    }

    fn run(sales: f64, var: f64, fixed: f64, interest: f64) -> LeverageResult {
        analyze(&LeverageInput {
            sales_usd: sales,
            variable_costs_usd: var,
            fixed_costs_usd: fixed,
            interest_expense_usd: interest,
        })
    }

    #[test]
    fn contribution_margin_and_ebit() {
        let r = run(1000.0, 400.0, 300.0, 100.0);
        assert!(close(r.contribution_margin_usd, 600.0));
        assert!(close(r.ebit_usd, 300.0));
        assert!(close(r.pretax_income_usd, 200.0));
    }

    #[test]
    fn dol() {
        // 600 / 300 = 2.0.
        assert!(close(run(1000.0, 400.0, 300.0, 100.0).dol.unwrap(), 2.0));
    }

    #[test]
    fn dfl() {
        // 300 / 200 = 1.5.
        assert!(close(run(1000.0, 400.0, 300.0, 100.0).dfl.unwrap(), 1.5));
    }

    #[test]
    fn dcl_is_product_and_cm_over_pretax() {
        let r = run(1000.0, 400.0, 300.0, 100.0);
        assert!(close(r.dcl.unwrap(), 3.0));
        assert!(close(r.dcl.unwrap(), r.dol.unwrap() * r.dfl.unwrap()));
        assert!(close(r.dcl.unwrap(), r.contribution_margin_usd / r.pretax_income_usd));
    }

    #[test]
    fn no_fixed_costs_no_interest_is_unit_leverage() {
        // All-variable, no fixed costs, no debt → no amplification.
        let r = run(1000.0, 400.0, 0.0, 0.0);
        assert!(close(r.dol.unwrap(), 1.0));
        assert!(close(r.dfl.unwrap(), 1.0));
        assert!(close(r.dcl.unwrap(), 1.0));
    }

    #[test]
    fn zero_ebit_guards_dol() {
        // CM exactly covers fixed costs → EBIT 0.
        let r = run(1000.0, 400.0, 600.0, 0.0);
        assert!(close(r.ebit_usd, 0.0));
        assert!(r.dol.is_none());
    }

    #[test]
    fn breakeven_pretax_guards_dfl_dcl() {
        // EBIT equals interest → pre-tax income 0.
        let r = run(1000.0, 400.0, 300.0, 300.0);
        assert!(close(r.pretax_income_usd, 0.0));
        assert!(r.dfl.is_none());
        assert!(r.dcl.is_none());
    }

    #[test]
    fn higher_fixed_costs_raise_dol() {
        let low = run(1000.0, 400.0, 100.0, 0.0);
        let high = run(1000.0, 400.0, 400.0, 0.0);
        assert!(high.dol.unwrap() > low.dol.unwrap());
    }
}
